File: Server/aoRss/Views/AggregatorClass.vb




Imports Contensive.Addons.Rss
Imports Contensive.Addons.Rss.Controllers
Imports Contensive.Addons.Rss.Models.Db
Imports Contensive.Addons.Rss.Models.Domain
Imports Contensive.BaseClasses

Namespace Views
    '
    Public Class AggregatorClass
        Inherits AddonBaseClass
        '
        '=====================================================================================
        ''' <summary>
        ''' AddonDescription
        ''' </summary>
        ''' <param name="CP"></param>
        ''' <returns></returns>
        Public Overrides Function Execute(CP As CPBaseClass) As Object
            Dim result As String = ""
            Dim sw As New Stopwatch : sw.Start()
            Try
                Dim AggregatorName As String = ""
                Dim cr1 As String = vbCrLf & vbTab
                Dim cr2 As String = cr1 & vbTab
                Dim cr3 As String = cr2 & vbTab
                Dim cr4 As String = cr3 & vbTab
                '
                Dim instanceId As String = CP.Doc.GetText("instanceId")
                CP.Utils.AppendLog("First Instance ID=" & instanceId)
                Dim StoryCnt As Long = CP.Utils.EncodeInteger(CP.Doc.GetText("Story count"))
                If StoryCnt = 0 Then
                    StoryCnt = 5
                End If
                Dim RSSAggregator As RSSAggregatorModel = BaseModel.create(Of RSSAggregatorModel)(CP, (instanceId))
                Dim AggregatorId As Integer = 0
                'cs = csv.opencsContent("RSS aggregators", "(name=" & CP.Utils.EncodeText(AggregatorName) & ")")
                If (RSSAggregator IsNot Nothing) Then
                    AggregatorId = RSSAggregator.id
                End If
                CP.Utils.AppendLog("Aggregator ID=" & AggregatorId)
                If (AggregatorId = 0) Or (instanceId Is Nothing) Then
                    '
                    ' Create new aggregator
                    '
                    RSSAggregator = BaseModel.add(Of RSSAggregatorModel)(CP)
                    RSSAggregator.name = "New RSS Aggregator create on-demand " & Now
                    AggregatorId = RSSAggregator.id
                    RSSAggregator.ccguid = instanceId
                    RSSAggregator.save(Of RSSAggregatorModel)(CP)
                    CP.Utils.AppendLog("instanceId=" & instanceId)

                    'End If
                End If
                '
                Dim cacheName As String = "rssAggregator:" & AggregatorId & ":" & StoryCnt
                Execute = CP.Doc.GetText(cacheName)
                If True Then
                    '
                    '
                    '
                    Dim storyList As List(Of RSSAggregatorSourceStorieModel) = RSSAggregatorSourceStorieModel.createStoryList(CP, CInt(AggregatorId))
                    '
                    Dim Ptr As Long = 1
                    Dim list As String = ""
                    For Each story In storyList
                        If StoryCnt >= Ptr Then
                            Dim Cell As String = ""
                            '
                            Dim Link As String = story.link
                            Dim Copy As String = Trim(story.name)
                            If Copy <> "" Then
                                If Link <> "" Then
                                    Copy = "<a href=""" & Link & """>" & Copy & "</a>"
                                End If
                                Cell = Cell & cr3 & "<h3 class=""raCaption"">" & Copy & "</h3>"
                            End If
                            CP.Utils.AppendLog("copy=" & Copy)
                            '
                            Dim pubDate As String = CType(story.pubDate, String)
                            Dim sourceName As String = Trim(story.name)
                            '
                            Dim Delimiter As String = ""
                            If sourceName <> "" Then
                                sourceName = "<span class=""raSourceName"">" & sourceName & "</span>"
                            End If
                            If pubDate <> "" Then
                                pubDate = "<span  class=""raPubDate"">" & pubDate & "</span >"
                            End If
                            If (pubDate <> "") And (sourceName <> "") Then
                                Delimiter = "<span class=""raDelimiter"">|</span>"
                            End If
                            Copy = sourceName & Delimiter & pubDate
                            If Copy <> "" Then
                                Cell = Cell & cr3 & "<p class=""raByLine"">" & Copy & "</p>"
                            End If
                            '
                            Copy = Trim(story.description)
                            If Copy <> "" Then
                                Cell = Cell & cr3 & "<p class=""raDescription"">" & Copy & "</p>"
                                ' Cell = Cell & cr3 & "<p class=""raDescription"">" & Copy & "</p><p><a href=""" & Link & """>Continue Reading</a></p>"
                            End If
                            '
                            list = list _
                                & cr2 & "<ul><li class=""raItem"">" _
                                & Cell _
                                & cr2 & "</li></ul>"
                            '

                            story.save(Of RSSAggregatorSourceStorieModel)(CP)
                            Ptr = Ptr + 1
                        Else
                            Exit For
                        End If

                    Next
                    If list <> "" Then
                        Execute = "" _
                            & cr1 & "<ul class=""rssAggregator"">" _
                            & list _
                            & cr1 & "</ul>" _
                            & ""
                    End If
                    result = list
                End If
            Catch ex As Exception
                CP.Site.ErrorReport(ex)
            End Try
            Return result
        End Function
        '
        '=====================================================================================
        '
        Friend Function GetXMLAttribute(cp As CPBaseClass, Found As Boolean, Node As Xml.XmlNode, Name As String) As String
            Dim result As String = ""
            Try
                Dim NodeAttribute As Xml.XmlAttribute
                Dim REsultNode As Xml.XmlNode
                Dim UcaseName As String
                Found = False
                REsultNode = Node.Attributes.GetNamedItem(Name)
                If (REsultNode Is Nothing) Then
                    UcaseName = UCase(Name)
                    For Each NodeAttribute In Node.Attributes
                        If UCase(NodeAttribute.Name) = UcaseName Then
                            GetXMLAttribute = NodeAttribute.Value
                            Found = True
                            Exit For
                        End If
                    Next
                Else
                    GetXMLAttribute = REsultNode.Value
                    Found = True
                End If
            Catch ex As Exception
                cp.Site.ErrorReport(ex)
            End Try
            Return result
        End Function
    End Class
    '
    '=====================================================================================
    '
    Friend Class EnclosureType
        Friend URL As String
        Public Property Type As String
        Public Property Length As String
    End Class
    '
End Namespace

File: Server/aoRss/Views/AggregatorClass.cs
using System;
using System.Diagnostics;
using Contensive.Addons.Rss.Models.Db;
using Contensive.BaseClasses;
using Contensive.Models.Db;

namespace Contensive.Addons.Rss.Views {
    //
    public class AggregatorClass : AddonBaseClass {
        //
        // =====================================================================================
        /// <summary>
        /// AddonDescription
        /// </summary>
        /// <param name="CP"></param>
        /// <returns></returns>
        public override object Execute(CPBaseClass CP) {
            object ExecuteRet = default;
            string result = "";
            var sw = new Stopwatch();
            sw.Start();
            try {
                string AggregatorName = "";
                string cr1 = "\r\n\t";
                string cr2 = cr1 + "\t";
                string cr3 = cr2 + "\t";
                string cr4 = cr3 + "\t";
                //
                string instanceId = CP.Doc.GetText("instanceId");
                CP.Utils.AppendLog($"First Instance ID={instanceId}");
                long StoryCnt = CP.Utils.EncodeInteger(CP.Doc.GetText("Story count"));
                if (StoryCnt == 0L) {
                    StoryCnt = 5L;
                }
                var RSSAggregator = DbBaseModel.create<RSSAggregatorModel>(CP, instanceId);
                int AggregatorId = 0;
                if (RSSAggregator is not null) {
                    AggregatorId = RSSAggregator.id;
                }
                CP.Utils.AppendLog($"Aggregator ID={AggregatorId}");
                if (AggregatorId == 0 | instanceId is null) {
                    //
                    // Create new aggregator
                    //
                    RSSAggregator = DbBaseModel.addDefault<RSSAggregatorModel>(CP);
                    RSSAggregator.name = $"New RSS Aggregator create on-demand {DateTime.Now}";
                    AggregatorId = RSSAggregator.id;
                    RSSAggregator.ccguid = instanceId;
                    RSSAggregator.save(CP);
                    CP.Utils.AppendLog($"instanceId={instanceId}");
                }
                //
                string cacheName = $"rssAggregator:{AggregatorId}:{StoryCnt}";
                ExecuteRet = CP.Doc.GetText(cacheName);
                if (true) {
                    //
                    //
                    //
                    var storyList = RSSAggregatorSourceStorieModel.createStoryList(CP, AggregatorId);
                    //
                    long Ptr = 1L;
                    string list = "";
                    foreach (var story in storyList) {
                        if (StoryCnt >= Ptr) {
                            string Cell = "";
                            //
                            string Link = story.link;
                            string Copy = (story.name ?? "").Trim();
                            if (!string.IsNullOrEmpty(Copy)) {
                                if (!string.IsNullOrEmpty(Link)) {
                                    Copy = $"<a href=\"{Link}\">{Copy}</a>";
                                }
                                Cell = $"{Cell}{cr3}<h3 class=\"raCaption\">{Copy}</h3>";
                            }
                            CP.Utils.AppendLog($"copy={Copy}");
                            //
                            string pubDate = story.pubDate.ToString();
                            string sourceName = (story.name ?? "").Trim();
                            //
                            string Delimiter = "";
                            if (!string.IsNullOrEmpty(sourceName)) {
                                sourceName = $"<span class=\"raSourceName\">{sourceName}</span>";
                            }
                            if (!string.IsNullOrEmpty(pubDate)) {
                                pubDate = $"<span  class=\"raPubDate\">{pubDate}</span >";
                            }
                            if (!string.IsNullOrEmpty(pubDate) & !string.IsNullOrEmpty(sourceName)) {
                                Delimiter = "<span class=\"raDelimiter\">|</span>";
                            }
                            Copy = sourceName + Delimiter + pubDate;
                            if (!string.IsNullOrEmpty(Copy)) {
                                Cell = $"{Cell}{cr3}<p class=\"raByLine\">{Copy}</p>";
                            }
                            //
                            Copy = (story.description ?? "").Trim();
                            if (!string.IsNullOrEmpty(Copy)) {
                                Cell = $"{Cell}{cr3}<p class=\"raDescription\">{Copy}</p>";
                            }
                            //
                            list = $"{list}{cr2}<ul><li class=\"raItem\">{Cell}{cr2}</li></ul>";

                            //

                            story.save(CP);
                            Ptr = Ptr + 1L;
                        } else {
                            break;
                        }

                    }
                    if (!string.IsNullOrEmpty(list)) {
                        ExecuteRet = $"{cr1}<ul class=\"rssAggregator\">{list}{cr1}</ul>";
                    }
                    result = list;
                }
            } catch (Exception ex) {
                CP.Site.ErrorReport(ex);
            }
            return result;
        }
        //
        // =====================================================================================
        //
        internal string GetXMLAttribute(CPBaseClass cp, bool Found, System.Xml.XmlNode Node, string Name) {
            string result = "";
            try {
                System.Xml.XmlNode resultNode;
                Found = false;
                if (Node.Attributes == null) {
                    return result;
                }
                resultNode = Node.Attributes.GetNamedItem(Name);
                if (resultNode is null) {
                    string UcaseName = Name.ToUpperInvariant();
                    foreach (System.Xml.XmlAttribute NodeAttribute in Node.Attributes) {
                        if (string.Equals(NodeAttribute.Name, UcaseName, StringComparison.OrdinalIgnoreCase)) {
                            result = NodeAttribute.Value;
                            Found = true;
                            break;
                        }
                    }
                } else {
                    result = resultNode.Value;
                    Found = true;
                }
            } catch (Exception ex) {
                cp.Site.ErrorReport(ex);
            }
            return result;
        }
    }
    //
    // =====================================================================================
    //
    internal class EnclosureType {
        internal string URL;
        public string Type { get; set; }
        public string Length { get; set; }
    }
    //
}
